### ISSUE TITLE
Release for Python 3.8.19, 3.9.19, and 3.10.14

### ISSUE THREAD

**Issue by indygreg** [indygreg/python-build-standalone#244]
CPython 3.8.19, 3.9.19, and 3.10.14 were released on March 19, 2024 and contain security fixes. We should upgrade and get a release out the door with these versions.

**Comment by zanieb** [comment]
I'd be interested in doing some work on this.

**Comment by charliermarsh** [comment]
I'll do the download bump now to start.

**Comment by indygreg** [comment]
@charliermarsh you are a collaborator on the repo now and the latest commit on `main` (`d09ff921d92d6da8d8a608eaa850dc8c0f638194`) should be appropriate for a release (assuming CI passes). You should be able to run `just release $(cat /path/to/github/token) d09ff921d92d6da8d8a608eaa850dc8c0f638194 20240414` to perform a release.

You have my blessing to run this command.

The release should be created as pre-release by default. Please make an attempt to populate the release notes. Then @ me here once done so I can look at things and we can promote the release.

If you run into release pipeline regressions during the release (not uncommon!), I usually just hack up code locally to get it to pass, push to `main` [via PR], then run the release automation from the fixed commit, targeting an older commit (so you don't have to wait for GitHub Actions to finish before releasing).

**Comment by charliermarsh** [comment]
Great, thanks @indygreg! I'll give this a try today.

**Comment by charliermarsh** [comment]
Is it possibly because the builds are all now re-running on that commit, since we created a tag for it?

**Comment by indygreg** [comment]
The Rust release automation is only supposed to look at CI invocations that are done. So multiple CI runs in the same commit shouldn't be an issue. But this logic has historically been buggy.

Earlier log snippet indicates the CI artifact wasn't a zip file. Feels like an intermittent failure or we messed up artifact uploading.

**Comment by charliermarsh** [comment]
Wondering if there's something wrong with my creds because I'm seeing it consistently for different artifacts; and when I download them out-of-band, they seem to unzip without error (e.g., `cpython-3.11-aarch64-apple-darwin-debug`). Looking into it.

**Comment by charliermarsh** [comment]
(Creds would be odd though since I'd expect an earlier failure _and_ my creds successfully created the Release itself.)

**Comment by charliermarsh** [comment]
Okay yeah the fetched artifact on my machine (when fetched via `github.rs`) is just 158 bytes. Looking at it...

**Comment by charliermarsh** [comment]
Hah, ok, the contents are just:
```json
{"message":"You must have the actions scope to download artifacts.","documentation_url":"https://docs.github.com/rest/actions/artifacts#download-an-artifact"}
```

**Comment by charliermarsh** [comment]
Ok, I generated a fine-grained PAT with the Actions scope and now it's moving along nicely.

**Comment by charliermarsh** [comment]
(Looking into both of these.)

**Comment by charliermarsh** [comment]
Ok I believe I've debugged the `shared-*` issue. Just requires a small change to `release.rs`.

**Comment by charliermarsh** [comment]
Should be fixed in https://github.com/indygreg/python-build-standalone/pull/250. (Would you prefer to create a new release tag, or just release locally once that is merged, since it only affects the upload process and not the assets?)

**Comment by indygreg** [comment]
That's just weird. So you can create a release on the repo but not upload an artifact to it? That seems like an odd permissions rift.

I want to believe different credentials are being used. It's the Rust code failing, not `gh`. So it must have something to do with the PAT used by Rust/just.

**Comment by charliermarsh** [comment]
Let me investigate that hypothesis. I've also tried creating fine-grained PATs with all permissions, and using a classic PAT, all with the same result.

**Comment by charliermarsh** [comment]
It seems to be using the token I provided, as far as I can tell? It's the only request that doesn't go through `octobcrab` though and the only request that (IIUC) is lacking representation in the semantic API.

**Comment by charliermarsh** [comment]
(Will dig into this again tonight.)

**Comment by charliermarsh** [comment]
I got this to work by creating a class PAT with every permission enabled. I'll follow up later to figure out the exact necessary combination.

**Comment by charliermarsh** [comment]
@indygreg - Okay, success, I think? I've populated the release notes but left it in draft here: https://github.com/indygreg/python-build-standalone/releases/tag/20240415

**Comment by indygreg** [comment]
I finally have some time on my desktop tonight. So I'll be taking a pass on the release notes and will likely remove the pre-release label. Thank you for your patience figuring out how to release.

Maybe a good follow-up would be to move the release automation to a parameterized GitHub Action workflow so we don't have to bother with local tokens.

**Comment by indygreg** [comment]
I just published the 20240415. Thanks again for your persistence!

**Comment by charliermarsh** [comment]
No prob, thank you for promoting! I'll think too on ways we can streamline it.
